File: packages/go/qorechain/tx/evm_window_test.go
package tx

import (
	"errors"
	"fmt"
	"testing"
)

// The chain's own refusal texts, as they arrive over the EVM JSON-RPC lane.
const (
	chainNoWindowText = "account qor1qypqxpq9qcrsszg2pvxq6rs0zqg3yyc5tfman7 has no open EVM " +
		"authorisation window; open one with MsgOpenEVMWindow, signed on the Cosmos lane with " +
		"the account's post-quantum key"
	chainNoKeyText = "account qor1qypqxpq9qcrsszg2pvxq6rs0zqg3yyc5tfman7 has no registered " +
		"post-quantum key"
)

// TestClassifyEVMWindowFailureByCode covers the three pqc codes, including the
// two states code 28 carries.
func TestClassifyEVMWindowFailureByCode(t *testing.T) {
	cases := []struct {
		name      string
		code      uint32
		codespace string
		text      string
		want      EVMWindowIssue
	}{
		{"code 26", 26, "pqc", chainNoWindowText, EVMWindowIssueNoWindow},
		{"code 26 without text", 26, "pqc", "", EVMWindowIssueNoWindow},
		{"code 27", 27, "pqc", "EVM authorisation window exhausted", EVMWindowIssueExhausted},
		{"code 28 invalid window", 28, "pqc", "invalid EVM authorisation window", EVMWindowIssueInvalid},
		{"code 28 no key", 28, "pqc", chainNoKeyText, EVMWindowIssueNoPQCKey},
		{"code 21 is not a window issue", 21, "pqc", "hybrid PQC signature verification failed", EVMWindowIssueNone},
		{"another codespace, same number", 26, "sdk", "unpacking protobuf message failed", EVMWindowIssueNone},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := ClassifyEVMWindowFailure(tc.code, tc.codespace, tc.text); got != tc.want {
				t.Fatalf("got %q, want %q", got, tc.want)
			}
		})
	}
}

// TestClassifyEVMWindowErrorByText covers the JSON-RPC lane, where the refusal
// carries the chain's text and no codespace at all.
func TestClassifyEVMWindowErrorByText(t *testing.T) {
	cases := []struct {
		name string
		err  error
		want EVMWindowIssue
	}{
		{"chain no-window text", errors.New(chainNoWindowText), EVMWindowIssueNoWindow},
		{"chain no-key text", errors.New(chainNoKeyText), EVMWindowIssueNoPQCKey},
		{"wrapped no-window text", fmt.Errorf("broadcast failed: %w", errors.New(chainNoWindowText)), EVMWindowIssueNoWindow},
		{"exhausted text", errors.New("EVM authorisation window exhausted for account qor1abc"), EVMWindowIssueExhausted},
		{"expired text", errors.New("EVM authorisation window expired at height 6069908"), EVMWindowIssueInvalid},
		{"american spelling", errors.New("account qor1abc has no open EVM authorization window"), EVMWindowIssueNoWindow},
		{"unrelated error", errors.New("insufficient funds: 10uqor is smaller than 20uqor"), EVMWindowIssueNone},
		{"unrelated nonce error", errors.New("nonce too low: next nonce 3, tx nonce 2"), EVMWindowIssueNone},
		{"nil", nil, EVMWindowIssueNone},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := ClassifyEVMWindowError(tc.err)
			if got != tc.want {
				t.Fatalf("got %q, want %q", got, tc.want)
			}
			if want := tc.want != EVMWindowIssueNone; IsEVMWindowError(tc.err) != want {
				t.Fatalf("IsEVMWindowError = %v, want %v", !want, want)
			}
		})
	}
}

// TestClassifyEVMWindowErrorFromTxError classifies a decoded ABCI failure, the
// shape the Cosmos lane returns.
func TestClassifyEVMWindowErrorFromTxError(t *testing.T) {
	err := DecodeTxError(26, "pqc", chainNoWindowText)
	if err == nil {
		t.Fatal("DecodeTxError returned nil for a non-zero code")
	}
	if got := ClassifyEVMWindowError(err); got != EVMWindowIssueNoWindow {
		t.Fatalf("got %q, want %q", got, EVMWindowIssueNoWindow)
	}
	if err.Reason != "no open EVM authorisation window" {
		t.Fatalf("unexpected reason: %q", err.Reason)
	}
	wrapped := fmt.Errorf("broadcast: %w", err)
	if got := ClassifyEVMWindowError(wrapped); got != EVMWindowIssueNoWindow {
		t.Fatalf("wrapped: got %q, want %q", got, EVMWindowIssueNoWindow)
	}

	keyErr := DecodeTxError(28, "pqc", chainNoKeyText)
	if got := ClassifyEVMWindowError(keyErr); got != EVMWindowIssueNoPQCKey {
		t.Fatalf("code 28 with the no-key text: got %q", got)
	}
	if got := ClassifyEVMWindowError(DecodeTxError(27, "pqc", "")); got != EVMWindowIssueExhausted {
		t.Fatalf("code 27: got %q", got)
	}
	if got := ClassifyEVMWindowError(DecodeTxError(5, "sdk", "insufficient funds")); got != EVMWindowIssueNone {
		t.Fatalf("unrelated tx error: got %q", got)
	}
}

// TestEVMWindowRemedies checks each issue names its own remedy and that the
// no-key remedy is the register-key one, not the open-a-window one.
func TestEVMWindowRemedies(t *testing.T) {
	for _, issue := range []EVMWindowIssue{
		EVMWindowIssueNoWindow, EVMWindowIssueExhausted, EVMWindowIssueInvalid, EVMWindowIssueNoPQCKey,
	} {
		if issue.Remedy() == "" {
			t.Fatalf("issue %q has no remedy", issue)
		}
	}
	if EVMWindowIssueNone.Remedy() != "" {
		t.Fatal("EVMWindowIssueNone must have no remedy")
	}
	if got := EVMWindowIssueNoPQCKey.Remedy(); got == EVMWindowIssueNoWindow.Remedy() {
		t.Fatal("the missing-key remedy must differ from the missing-window remedy")
	}
}

File: packages/java/src/test/java/io/github/qorechain/EvmWindowTest.java
package io.github.qorechain;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.google.protobuf.Any;
import io.github.qorechain.evm.EvmWindow;
import io.github.qorechain.evm.EvmWindowIssue;
import io.github.qorechain.evm.EvmWindowStatus;
import io.github.qorechain.messages.Messages;
import io.github.qorechain.messages.TypedMessage;
import io.github.qorechain.tx.TxError;
import java.math.BigInteger;
import org.junit.jupiter.api.Test;

/**
 * The EVM authorisation window (chain v3.2.0): composer round-trip, the
 * ValidateBasic bounds, both query shapes and the error classifier.
 */
class EvmWindowTest {

    private static final String SENDER = "qor1qypqxpq9qcrsszg2pvxq6rs0zqg3yyc5tfman7";

    private static final String CHAIN_NO_WINDOW_TEXT =
            "account "
                    + SENDER
                    + " has no open EVM authorisation window; open one with MsgOpenEVMWindow,"
                    + " signed on the Cosmos lane with the account's post-quantum key";

    private static final String CHAIN_NO_KEY_TEXT =
            "account " + SENDER + " has no registered post-quantum key";

    // ---- composers ----

    @Test
    void openWindowRoundTripsThroughTheRegistry() throws Exception {
        TypedMessage tm = EvmWindow.open(SENDER, 300, 5, "2000000");
        assertEquals("/qorechain.pqc.v1.MsgOpenEVMWindow", tm.typeUrl);

        Any any = Messages.pack(tm);
        assertEquals("/qorechain.pqc.v1.MsgOpenEVMWindow", any.getTypeUrl());
        assertEquals(tm.message.toByteString(), any.getValue());

        qorechain.pqc.v1.Tx.MsgOpenEVMWindow decoded =
                (qorechain.pqc.v1.Tx.MsgOpenEVMWindow) Messages.unpack(any);
        assertEquals(SENDER, decoded.getSender());
        assertEquals(300L, decoded.getBlocks());
        assertEquals(5L, decoded.getMaxTxs());
        assertEquals("2000000", decoded.getMaxValue());
    }

    @Test
    void closeWindowRoundTripsThroughTheRegistry() throws Exception {
        TypedMessage tm = EvmWindow.close(SENDER);
        assertEquals("/qorechain.pqc.v1.MsgCloseEVMWindow", tm.typeUrl);

        Any any = Messages.pack(tm);
        assertEquals(tm.message.toByteString(), any.getValue());
        qorechain.pqc.v1.Tx.MsgCloseEVMWindow decoded =
                (qorechain.pqc.v1.Tx.MsgCloseEVMWindow) Messages.unpack(any);
        assertEquals(SENDER, decoded.getSender());
    }

    // ---- validation ----

    @Test
    void boundsAreRefusedWithAMessageNamingTheBound() {
        assertTrue(refusal(0, 5, "1000").contains("blocks must be between 1 and 17280"));
        assertTrue(refusal(17281, 5, "1000").contains("blocks must be between 1 and 17280"));
        assertTrue(refusal(300, 0, "1000").contains("max_txs must be between 1 and 1000"));
        assertTrue(refusal(300, 1001, "1000").contains("max_txs must be between 1 and 1000"));
        assertTrue(refusal(300, 5, "0").contains("max_value must be greater than 0 uqor"));
        assertTrue(refusal(300, 5, "-1").contains("max_value must be greater than 0 uqor"));
        assertTrue(refusal(300, 5, "").contains("max_value is required"));
        assertTrue(refusal(300, 5, "1.5").contains("not an integer amount of uqor"));
    }

    @Test
    void boundsThemselvesAreAccepted() {
        assertEquals(1L, EvmWindow.openMessage(SENDER, 1, 1, "1").getBlocks());
        qorechain.pqc.v1.Tx.MsgOpenEVMWindow max =
                EvmWindow.openMessage(SENDER, EvmWindow.MAX_BLOCKS, EvmWindow.MAX_TXS, "1");
        assertEquals(17280L, max.getBlocks());
        assertEquals(1000L, max.getMaxTxs());
    }

    @Test
    void senderIsRequiredAndMustBeAQorAddress() {
        assertTrue(
                assertThrows(IllegalArgumentException.class, () -> EvmWindow.open("", 300, 5, "1000"))
                        .getMessage()
                        .contains("sender is required"));
        assertTrue(
                assertThrows(
                                IllegalArgumentException.class,
                                () -> EvmWindow.open("qor1nope", 300, 5, "1000"))
                        .getMessage()
                        .contains("not a valid qor bech32 address"));
        assertTrue(
                assertThrows(IllegalArgumentException.class, () -> EvmWindow.close("cosmos1xyz"))
                        .getMessage()
                        .contains("not a valid qor bech32 address"));
    }

    private static String refusal(long blocks, long maxTxs, String maxValue) {
        return assertThrows(
                        IllegalArgumentException.class,
                        () -> EvmWindow.open(SENDER, blocks, maxTxs, maxValue))
                .getMessage();
    }

    // ---- query ----

    @Test
    void parsesTheAbsentShape() {
        EvmWindowStatus s = EvmWindowStatus.parse("{\"found\":false}");
        assertFalse(s.found);
        assertFalse(s.live);
        assertFalse(s.exhausted(), "an absent window is not an exhausted window");
        assertEquals(BigInteger.ZERO, s.maxValue);
        assertEquals(BigInteger.ZERO, s.usedValue);
        assertEquals(BigInteger.ZERO, s.remainingValue);
        assertEquals(0L, s.remainingTxs);
    }

    @Test
    void parsesTheLiveShape() {
        EvmWindowStatus s =
                EvmWindowStatus.parse(
                        "{\"found\":true,\"live\":true,\"opened_height\":\"6069608\","
                                + "\"expiry_height\":\"6069908\",\"max_txs\":\"5\",\"used_txs\":\"1\","
                                + "\"max_value\":\"2000000\",\"used_value\":\"3363\","
                                + "\"remaining_blocks\":\"286\",\"remaining_txs\":\"4\","
                                + "\"remaining_value\":\"1996637\"}");
        assertTrue(s.found);
        assertTrue(s.live);
        assertEquals(6069608L, s.openedHeight);
        assertEquals(6069908L, s.expiryHeight);
        assertEquals(5L, s.maxTxs);
        assertEquals(1L, s.usedTxs);
        assertEquals(new BigInteger("2000000"), s.maxValue);
        assertEquals(new BigInteger("3363"), s.usedValue);
        assertEquals(286L, s.remainingBlocks);
        assertEquals(4L, s.remainingTxs);
        assertEquals(new BigInteger("1996637"), s.remainingValue);
        assertFalse(s.exhausted());
    }

    @Test
    void keepsBigNumbersExact() {
        // 2^53 + 1 and a value far beyond any 64-bit type: a float64 decode would
        // truncate both.
        EvmWindowStatus s =
                EvmWindowStatus.parse(
                        "{\"found\":true,\"live\":false,\"opened_height\":\"9007199254740993\","
                                + "\"expiry_height\":\"9007199254740995\",\"max_txs\":\"1000\","
                                + "\"used_txs\":\"1000\",\"max_value\":\"123456789012345678901234567890\","
                                + "\"used_value\":\"123456789012345678901234567890\","
                                + "\"remaining_blocks\":\"2\",\"remaining_txs\":\"0\","
                                + "\"remaining_value\":\"0\"}");
        assertEquals(9007199254740993L, s.openedHeight);
        assertEquals(9007199254740995L, s.expiryHeight);
        assertEquals(new BigInteger("123456789012345678901234567890"), s.maxValue);
        assertEquals(new BigInteger("123456789012345678901234567890"), s.usedValue);
        assertTrue(s.exhausted(), "a found-but-not-live window is exhausted");
    }

    @Test
    void acceptsBareNumbersAndRefusesGarbage() {
        EvmWindowStatus s =
                EvmWindowStatus.parse("{\"found\":true,\"live\":true,\"max_txs\":5,\"max_value\":2000000}");
        assertEquals(5L, s.maxTxs);
        assertEquals(new BigInteger("2000000"), s.maxValue);

        assertThrows(
                IllegalArgumentException.class,
                () -> EvmWindowStatus.parse("{\"found\":true,\"max_value\":\"1.5\"}"));
        assertThrows(
                IllegalArgumentException.class,
                () -> EvmWindowStatus.parse("{\"found\":true,\"max_txs\":\"many\"}"));
    }

    @Test
    void adaptsTheProtobufAnswer() {
        EvmWindowStatus s =
                EvmWindowStatus.of(
                        qorechain.pqc.v1.QueryOuterClass.QueryEVMWindowResponse.newBuilder()
                                .setFound(true)
                                .setLive(true)
                                .setMaxTxs(5)
                                .setMaxValue("2000000")
                                .setRemainingValue("1996637")
                                .build());
        assertTrue(s.found);
        assertEquals(new BigInteger("1996637"), s.remainingValue);
        // An unset cosmos.Int string is zero, not a parse failure.
        assertEquals(BigInteger.ZERO, s.usedValue);
    }

    @Test
    void restPathIsTheDocumentedRoute() {
        assertEquals("/qorechain/pqc/v1/evm_window/" + SENDER, EvmWindow.restPath(SENDER));
    }

    // ---- error classification ----

    @Test
    void classifiesTheThreePqcCodes() {
        assertEquals(
                EvmWindowIssue.NO_WINDOW, EvmWindowIssue.classify(26, "pqc", CHAIN_NO_WINDOW_TEXT));
        assertEquals(EvmWindowIssue.NO_WINDOW, EvmWindowIssue.classify(26, "pqc", ""));
        assertEquals(
                EvmWindowIssue.EXHAUSTED,
                EvmWindowIssue.classify(27, "pqc", "EVM authorisation window exhausted"));
        assertEquals(
                EvmWindowIssue.INVALID,
                EvmWindowIssue.classify(28, "pqc", "invalid EVM authorisation window"));
        assertEquals(EvmWindowIssue.NO_PQC_KEY, EvmWindowIssue.classify(28, "pqc", CHAIN_NO_KEY_TEXT));
        // Not a window issue.
        assertEquals(
                EvmWindowIssue.NONE,
                EvmWindowIssue.classify(21, "pqc", "hybrid PQC signature verification failed"));
        assertEquals(
                EvmWindowIssue.NONE,
                EvmWindowIssue.classify(26, "sdk", "unpacking protobuf message failed"));
    }

    @Test
    void classifiesTheChainTextsOverJsonRpc() {
        assertEquals(
                EvmWindowIssue.NO_WINDOW,
                EvmWindowIssue.classify(new RuntimeException(CHAIN_NO_WINDOW_TEXT)));
        assertEquals(
                EvmWindowIssue.NO_PQC_KEY, EvmWindowIssue.classify(new RuntimeException(CHAIN_NO_KEY_TEXT)));
        assertEquals(
                EvmWindowIssue.NO_WINDOW,
                EvmWindowIssue.classify(
                        new RuntimeException("broadcast failed", new RuntimeException(CHAIN_NO_WINDOW_TEXT))));
        assertEquals(
                EvmWindowIssue.EXHAUSTED,
                EvmWindowIssue.classify(new RuntimeException("EVM authorisation window exhausted")));
        assertEquals(
                EvmWindowIssue.INVALID,
                EvmWindowIssue.classify(
                        new RuntimeException("EVM authorisation window expired at height 6069908")));
        assertEquals(
                EvmWindowIssue.NO_WINDOW,
                EvmWindowIssue.classify(
                        new RuntimeException("account qor1abc has no open EVM authorization window")));
    }

    @Test
    void doesNotMatchUnrelatedErrors() {
        assertEquals(
                EvmWindowIssue.NONE,
                EvmWindowIssue.classify(new RuntimeException("insufficient funds: 10uqor < 20uqor")));
        assertEquals(
                EvmWindowIssue.NONE,
                EvmWindowIssue.classify(new RuntimeException("nonce too low: next nonce 3, tx nonce 2")));
        assertEquals(EvmWindowIssue.NONE, EvmWindowIssue.classify((Throwable) null));
        assertFalse(EvmWindowIssue.isWindowError(new RuntimeException("out of gas")));
    }

    @Test
    void classifiesADecodedAbciFailure() {
        TxError.QoreTxException err = TxError.decode(26, "pqc", CHAIN_NO_WINDOW_TEXT, "ABC123");
        assertEquals("evm_window_missing", err.kind);
        assertEquals(EvmWindowIssue.NO_WINDOW, EvmWindowIssue.classify(err));
        assertTrue(EvmWindowIssue.isWindowError(err));

        assertEquals(
                EvmWindowIssue.NO_PQC_KEY,
                EvmWindowIssue.classify(TxError.decode(28, "pqc", CHAIN_NO_KEY_TEXT, "")));
        assertEquals(
                EvmWindowIssue.EXHAUSTED, EvmWindowIssue.classify(TxError.decode(27, "pqc", "", "")));
        assertEquals(
                EvmWindowIssue.NONE,
                EvmWindowIssue.classify(TxError.decode(5, "sdk", "insufficient funds", "")));
    }

    @Test
    void everyIssueNamesItsOwnRemedy() {
        for (EvmWindowIssue issue :
                new EvmWindowIssue[] {
                    EvmWindowIssue.NO_WINDOW,
                    EvmWindowIssue.EXHAUSTED,
                    EvmWindowIssue.INVALID,
                    EvmWindowIssue.NO_PQC_KEY
                }) {
            assertFalse(issue.remedy().isEmpty(), issue + " has no remedy");
        }
        assertEquals("", EvmWindowIssue.NONE.remedy());
        assertNotEquals(EvmWindowIssue.NO_WINDOW.remedy(), EvmWindowIssue.NO_PQC_KEY.remedy());
    }
}
